File: frontend/src/services/collegeService.ts
import { api } from './api';
import type {
  College,
  Department,
  Program,
  AcademicYear,
  Semester,
  Section,
  ServiceStatus,
} from '../types';

export const collegeService = {
  async getStatus(): Promise<ServiceStatus> {
    const res = await api.get('/api/colleges/status');
    return res.data;
  },

  async getAllColleges(): Promise<College[]> {
    const res = await api.get('/api/colleges');
    return res.data;
  },

  async getCollegeById(id: number): Promise<College> {
    const res = await api.get(`/api/colleges/${id}`);
    return res.data;
  },

  async createCollege(data: Partial<College>): Promise<College> {
    const res = await api.post('/api/colleges', data);
    return res.data;
  },

  // Departments
  async getDepartments(collegeId: number): Promise<Department[]> {
    const res = await api.get(`/api/colleges/${collegeId}/departments`);
    return res.data;
  },

  async createDepartment(collegeId: number, data: Partial<Department>): Promise<Department> {
    const res = await api.post(`/api/colleges/${collegeId}/departments`, data);
    return res.data;
  },

  // Programs
  async getPrograms(collegeId: number): Promise<Program[]> {
    const res = await api.get(`/api/colleges/${collegeId}/programs`);
    return res.data;
  },

  async createProgram(collegeId: number, data: Partial<Program>): Promise<Program> {
    const res = await api.post(`/api/colleges/${collegeId}/programs`, data);
    return res.data;
  },

  // Academic Years
  async getAcademicYears(collegeId: number): Promise<AcademicYear[]> {
    const res = await api.get(`/api/colleges/${collegeId}/academic-years`);
    return res.data;
  },

  async createAcademicYear(collegeId: number, data: Partial<AcademicYear>): Promise<AcademicYear> {
    const res = await api.post(`/api/colleges/${collegeId}/academic-years`, data);
    return res.data;
  },

  // Semesters
  async getSemesters(collegeId: number): Promise<Semester[]> {
    const res = await api.get(`/api/colleges/${collegeId}/semesters`);
    return res.data;
  },

  async createSemester(collegeId: number, data: Partial<Semester>): Promise<Semester> {
    const res = await api.post(`/api/colleges/${collegeId}/semesters`, data);
    return res.data;
  },

  // Sections
  async getSections(collegeId: number): Promise<Section[]> {
    const res = await api.get(`/api/colleges/${collegeId}/sections`);
    return res.data;
  },

  async createSection(collegeId: number, data: Partial<Section>): Promise<Section> {
    const res = await api.post(`/api/colleges/${collegeId}/sections`, data);
    return res.data;
  },
};

File: backend/college-service/src/main/java/com/smartcampus/college/service/CollegeService.java
package com.smartcampus.college.service;

import com.smartcampus.college.entity.*;
import com.smartcampus.college.repository.*;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class CollegeService {

    private final CollegeRepository collegeRepository;
    private final DepartmentRepository departmentRepository;
    private final ProgramRepository programRepository;
    private final AcademicYearRepository academicYearRepository;
    private final SemesterRepository semesterRepository;
    private final SectionRepository sectionRepository;

    // College Operations
    public College createCollege(College college) {
        if (collegeRepository.existsByCollegeCode(college.getCollegeCode())) {
            throw new IllegalArgumentException("College code already exists: " + college.getCollegeCode());
        }
        return collegeRepository.save(college);
    }

    public List<College> getAllColleges() {
        return collegeRepository.findAll();
    }

    public College getCollegeById(Long id) {
        return collegeRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("College not found with id: " + id));
    }

    // Department Operations
    public Department createDepartment(Department department) {
        getCollegeById(department.getCollegeId()); // Ensure college exists
        return departmentRepository.save(department);
    }

    public List<Department> getDepartmentsByCollege(Long collegeId) {
        return departmentRepository.findByCollegeId(collegeId);
    }

    // Program Operations
    public Program createProgram(Program program) {
        getCollegeById(program.getCollegeId());
        return programRepository.save(program);
    }

    public List<Program> getProgramsByCollege(Long collegeId) {
        return programRepository.findByCollegeId(collegeId);
    }

    // Academic Year Operations
    public AcademicYear createAcademicYear(AcademicYear academicYear) {
        getCollegeById(academicYear.getCollegeId());
        return academicYearRepository.save(academicYear);
    }

    public List<AcademicYear> getAcademicYearsByCollege(Long collegeId) {
        return academicYearRepository.findByCollegeId(collegeId);
    }

    // Semester Operations
    public Semester createSemester(Semester semester) {
        getCollegeById(semester.getCollegeId());
        return semesterRepository.save(semester);
    }

    public List<Semester> getSemestersByCollege(Long collegeId) {
        return semesterRepository.findByCollegeId(collegeId);
    }

    // Section Operations
    public Section createSection(Section section) {
        getCollegeById(section.getCollegeId());
        return sectionRepository.save(section);
    }

    public List<Section> getSectionsByCollege(Long collegeId) {
        return sectionRepository.findByCollegeId(collegeId);
    }
}
